core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala

```diff
@@ -438,12 +439,27 @@ class BlockManagerMasterEndpoint(
     if (blockLocations.containsKey(blockId)) blockLocations.get(blockId).toSeq else Seq.empty
   }
 
-  private def getLocationsAndStatus(blockId: BlockId): Option[BlockLocationsAndStatus] = {
+  private def getLocationsAndStatus(
+      blockId: BlockId, requesterHost: String): Option[BlockLocationsAndStatus] = {
     val locations = Option(blockLocations.get(blockId)).map(_.toSeq).getOrElse(Seq.empty)
     val status = locations.headOption.flatMap { bmId => blockManagerInfo(bmId).getStatus(blockId) }
 
     if (locations.nonEmpty && status.isDefined) {
-      Some(BlockLocationsAndStatus(locations, status.get))
+      val bmIdToLocalDirs = if (status.get.storageLevel.useDisk) {
+        locations
+          .find(_.host == requesterHost)
+          .map { sameHostBlockId =>
+            val bmInfo = blockManagerInfo(sameHostBlockId)
+            bmInfo.blockManagerId -> bmInfo.localDirs
+          }
+      } else {
+        None
+      }
+      bmIdToLocalDirs.map { case (bmId, localDirs) =>
+        BlockLocationsAndStatus(locations.filter(_ != bmId), status.get, Some(localDirs))
```

Comment:
Hmm, this looks a little counter-intuitive, so a comment could help...

I guess what you're doing here is sending back a list of block manager IDs where the block is stored, except for the one that owns the `localDirs` you're sending back?

Will it break anything if you just send back the list of locations untouched?

Reply:
I was thinking about this case and you are right better not use filterer out this one item from the locations: although this method is only used by one RPC call `GetLocationsAndStatus` which only used by the `BlockManager#getRemoteManagedBuffer` where the `localDirs` are used for accessing the block content and if that fails still there is some chance that this could be loaded from Memory (if the block was in Memory then spilled to Disk then promoted to Memory latter on, like Imran mentioned in #24499 ).